pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaMessageBatch.java

```diff
@@ -71,52 +61,22 @@ public int getUnfilteredMessageCount() {
   }
 
   @Override
-  public StreamMessage getMessageAtIndex(int index) {
-    return _messageList.get(index);
-  }
-
-  @Override
-  public int getMessageOffsetAtIndex(int index) {
-    return ByteBuffer.wrap(_messageList.get(index).getValue()).arrayOffset();
-  }
-
-  @Override
-  public int getMessageLengthAtIndex(int index) {
-    return _messageList.get(index).getValue().length;
-  }
-
-  @Override
-  public long getNextStreamMessageOffsetAtIndex(int index) {
-    throw new UnsupportedOperationException("This method is deprecated");
-  }
-
-  @Override
-  public StreamPartitionMsgOffset getNextStreamPartitionMsgOffsetAtIndex(int index) {
-    return new LongMsgOffset(((KafkaStreamMessage) _messageList.get(index)).getNextOffset());
+  public BytesStreamMessage getStreamMessage(int index) {
+    return _messages.get(index);
   }
 
   @Override
   public StreamPartitionMsgOffset getOffsetOfNextBatch() {
-    return new LongMsgOffset(_lastOffset + 1);
-  }
-
-  @Override
-  public RowMetadata getMetadataAtIndex(int index) {
-    return _messageList.get(index).getMetadata();
+    return new LongMsgOffset(_offsetOfNextBatch);
   }
 
   @Override
-  public byte[] getMessageBytesAtIndex(int index) {
-    return _messageList.get(index).getValue();
-  }
-
-  @Override
-  public StreamMessage getStreamMessage(int index) {
-    return _messageList.get(index);
+  public StreamPartitionMsgOffset getFirstMessageOffset() {
```

Comment:
We perform a null check after reading the value (the logic already exists), so this is actually a bug fix. When there is no value consumed, we should return `null` instead of negative offset. It is backward compatible.
Added an `@Nullable` annotation to this method